File: src/lb/perft.zig
fn core(board: *lb.Board, depth: usize) usize {
    if (depth == 0) return 1;
    var result: usize = 0;
    var moves = lb.MoveList{};
    moves.generateMoves(board);
    for (moves.moves.slice()) |m| {
        var new_board = board.*;
        new_board.move(m);
        result += core(&new_board, depth - 1);
    }
    return result;
}

pub fn go(out: anytype, board: *lb.Board, depth: usize) !usize {
    if (depth == 0) return 1;

    var result: usize = 0;
    var moves = lb.MoveList{};
    var timer = std.time.Timer.start() catch @panic("no timer available");
    moves.generateMoves(board);
    for (moves.moves.slice()) |m| {
        var new_board = board.*;
        new_board.move(m);
        const p = core(&new_board, depth - 1);
        result += p;
        try out.raw("{}: {}\n", .{ m, p });
        try out.flush();
    }
    const elapsed: f64 = @floatFromInt(timer.read());
    try out.raw("Nodes searched (depth {}): {}\n", .{ depth, result });
    try out.raw("Search completed in {d:.1}ms\n", .{elapsed / std.time.ns_per_ms});
    try out.flush();

    return result;
}

const std = @import("std");
const lb = @import("../lb.zig");
const util = @import("../util.zig");

File: src/lb/perft.cpp
#include "lb/perft.h"

#include <print>

#include "lb/common.h"
#include "lb/movegen.h"
#include "lb/position.h"
#include "lb/types.h"

namespace lb::perft {

  template <bool print> static auto core(const Position &position, usize depth) -> usize {
    if (depth == 0)
      return 1;

    usize result = 0;

    movegen::MoveList moves;
    movegen::generateMoves(moves, position);

    if (depth == 1 && !print)
      return moves.size();

    for (Move m : moves) {
      const Position new_position = position.move(m);
      const usize child = core<false>(new_position, depth - 1);
      if constexpr (print) {
        std::print("{}: {}\n", m, child);
      }
      result += child;
    }

    return result;
  }

  auto value(const Position &position, usize depth) -> usize { return core<false>(position, depth); }

  auto run(const Position &position, usize depth) -> void {
    const auto start = time::Clock::now();
    const usize total = core<true>(position, depth);
    const auto end = time::Clock::now();

    const time::FloatSeconds elapsed = end - start;

    std::print("total: {}\n", total);
    std::print("perft to depth {} complete in {:.1f}ms ({:.1f} Mnps)\n", depth, elapsed.count() * 1000, total / (1'000'000 * elapsed.count()));
  }

} // namespace lb::perft
